tests-web/assets/js/main.js

```diff
@@ -1,46 +1,76 @@
 import { makeDropHandler, makeLinkHandler, handleDragOver } from "./files.js";
 import { makeOpenTest, loadTests } from "./tests.js";
 
+// TODO The hash changes every time the Rust code does, this is very brittle.
+// See https://github.com/thedodd/trunk/issues/230 or stop using trunk.
+import { import_osm } from "../osm2streets-npm-54d7959f4e2ba1e6.js";
+
 const useMap = (map) => {
-    const container = map.getContainer();
-    container.ondrop = makeDropHandler(map);
-    container.ondragover = handleDragOver;
-
-    map.loadLink = makeLinkHandler(map);
-    map.openTest = makeOpenTest(map)
-    console.info("New map created! File drops enabled.", container);
-
-
-    // Here we read the test name from the URL.
-    const q = new URLSearchParams(window.location.search);
-    if (q.has('test')) {
-        const test = q.get('test');
-        console.info("Loading test " + test + " from URL.");
-        map.openTest(test);
-    }
-
-    loadTests();
-}
-
-// Smuggle a reference to the created map, so I can work with it in JS land.
-console.debug("Listening in on map creations...");
-const LM = L.Map;
-window.maps = [];
-L.Map = function(x, opts = {
-    maxZoom: 21,
-}, ...args) {
-    const m = new LM(x, opts, ...args)
-    window.maps.push(m);
-    setTimeout(() => useMap(m), 0);
-    return m;
-}
-
-// Settings for tile layers.
-const LL = L.TileLayer;
-L.TileLayer = function(x, opts = {
-    maxNativeZoom: 18,
-    maxZoom: 21,
-}, ...args) {
-    return new LL(x, opts, ...args);
-}
+  const container = map.getContainer();
+  container.ondrop = makeDropHandler(map);
+  container.ondragover = handleDragOver;
+
+  map.loadLink = makeLinkHandler(map);
+  map.openTest = makeOpenTest(map);
+  console.info("New map created! File drops enabled.", container);
+
+  // Here we read the test name from the URL.
+  const q = new URLSearchParams(window.location.search);
+  if (q.has("test")) {
+    const test = q.get("test");
+    console.info("Loading test " + test + " from URL.");
+    map.openTest(test);
+  }
+
+  loadTests();
+};
+
+// Initialize the map
+const map = L.map("map", { maxZoom: 21 }).setView([40.0, 10.0], 4);
+L.tileLayer("https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png", {
+  maxNativeZoom: 18,
+  maxZoom: 21,
+  attribution: "© OpenStreetMap",
+}).addTo(map);
+useMap(map);
+
+// TODO Should we split this into another .js file? How should it get the map if so?
+// TODO Is it OK to just assume the button exists by the time this script runs,
+// or should we create the button here?
```

Comment:
Yes, it should go into it's own file, and have some initialisation function that accepts the map that it should attach to, which is what I've done with the `makeX` functions. That pattern is probably not the best fit, I am thinking of replacing each of them with a class instead (because the "class" metaphor is exactly what we are trying to do).

The button can either be part of the initial HTML, in which case we can assume it exists already. Alternatively, the initialisation of this feature could create the button (and append it to a "toolbar" element), which encapsulates the feature nicely.

Don't worry about the overall structure yet, that is something that I have been working through in my head, that I intend to tackle soon.

Reply:
Done